keras_cv/layers/preprocessing3d/frustum_random_dropping_points.py

```diff
@@ -0,0 +1,105 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import tensorflow as tf
+
+from keras_cv.layers.preprocessing3d import base_augmentation_layer_3d
+from keras_cv.ops import point_cloud
+
+POINT_CLOUDS = base_augmentation_layer_3d.POINT_CLOUDS
+BOUNDING_BOXES = base_augmentation_layer_3d.BOUNDING_BOXES
+
+
+class FrustumRandomDroppingPoints(base_augmentation_layer_3d.BaseAugmentationLayer3D):
+    """A preprocessing layer which randomly drops point within a randomly generated frustum during training.
+
+    This layer will randomly select a point from the point cloud as the center of a frustum then generate a frustum based
+    on r_distance, theta_width, and phi_width. Points inside the selected frustum are randomly dropped based on keep_probability.
+    During inference time, the output will be identical to input. Call the layer with `training=True` to drop the input points.
+
+    Input shape:
+      point_clouds: 3D (multi frames) float32 Tensor with shape
+        [num of frames, num of points, num of point features].
+        The first 5 features are [x, y, z, class, range].
+      bounding_boxes: 3D (multi frames) float32 Tensor with shape
+        [num of frames, num of boxes, num of box features].
+        The first 7 features are [x, y, z, dx, dy, dz, phi].
+
+    Output shape:
+      A dictionary of Tensors with the same shape as input Tensors.
+
+    Arguments:
+      drop_rate: A float scalar sets the probability threshold for dropping the points.
+      r_distance: A float scalar sets the starting distance of a frustum.
+      theta_width: A float scalar sets the theta width of a frustum.
+      phi_width: A float scalar sets the phi width of a frustum.
+    """
+
+    def __init__(self, r_distance, theta_width, phi_width, drop_rate=None, **kwargs):
+        super().__init__(**kwargs)
+
+        if r_distance < 0:
+            raise ValueError("r_distance must be >=0.")
+        if theta_width < 0:
+            raise ValueError("theta_width must be >=0.")
+        if phi_width < 0:
+            raise ValueError("phi_width must be >=0.")
+        drop_rate = drop_rate if drop_rate else 0.0
+        if drop_rate > 1:
+            raise ValueError("drop_rate must be <=1.")
+
+        self._r_distance = r_distance
+        self._theta_width = theta_width
+        self._phi_width = phi_width
+        keep_probability = 1 - drop_rate
+        self._keep_probability = keep_probability
+
+    def get_random_transformation(self, point_clouds, **kwargs):
+        # Randomly select a point from the first frame as the center of the frustum.
+        num_valid_points = tf.cast(tf.math.reduce_sum(point_clouds[0, :, 3]), tf.int32)
```

Comment:
can you create a macro definition for 3?

Reply:
Maybe it is better to define a struct for point_clouds?

Reply:
Sure, I thought we would define a struct for point_clouds, just like what we did for bounding_boxes.

Reply:
e.g., https://github.com/keras-team/keras-cv/pull/1064